prov/efa/src/rxr/rxr_read.c

```diff
@@ -553,69 +633,33 @@ int rxr_read_post(struct rxr_ep *ep, struct rxr_read_entry *read_entry)
 		assert(rma_iov_idx < read_entry->rma_iov_count);
 		assert(rma_iov_offset < read_entry->rma_iov[rma_iov_idx].len);
 
-		iov.iov_base = (char *)read_entry->iov[iov_idx].iov_base + iov_offset;
-		iov.iov_len = read_entry->iov[iov_idx].iov_len - iov_offset;
-
-		rma_iov.addr = (uintptr_t)read_entry->rma_iov[rma_iov_idx].addr + rma_iov_offset;
-		rma_iov.len = read_entry->rma_iov[rma_iov_idx].len - rma_iov_offset;
-		rma_iov.key = read_entry->rma_iov[rma_iov_idx].key;
+		read_once_len = MIN(read_entry->iov[iov_idx].iov_len - iov_offset,
+				    read_entry->rma_iov[rma_iov_idx].len - rma_iov_offset);
 
-		iov.iov_len = MIN(iov.iov_len, rma_iov.len);
 		if (read_entry->lower_ep_type == EFA_EP)
-			iov.iov_len = MIN(iov.iov_len, rxr_env.efa_read_segment_size);
-		iov.iov_len = MIN(iov.iov_len, max_read_size);
-		rma_iov.len = iov.iov_len;
-
-		/* because fi_send uses a pkt_entry as context
-		 * we had to use a pkt_entry as context too
-		 */
-		if (read_entry->lower_ep_type == SHM_EP)
-			pkt_entry = rxr_pkt_entry_alloc(ep, ep->shm_tx_pkt_pool, RXR_PKT_FROM_SHM_TX_POOL);
-		else
-			pkt_entry = rxr_pkt_entry_alloc(ep, ep->efa_tx_pkt_pool, RXR_PKT_FROM_EFA_TX_POOL);
-
-		if (OFI_UNLIKELY(!pkt_entry))
-			return -FI_EAGAIN;
-
-		rxr_pkt_init_read_context(ep, read_entry, iov.iov_len, pkt_entry);
-
-		memset(&msg, 0, sizeof(msg));
-		msg.msg_iov = &iov;
-		msg.desc = &read_entry->mr_desc[iov_idx];
-		msg.iov_count = 1;
-		msg.rma_iov = &rma_iov;
-		msg.rma_iov_count = 1;
-		msg.context = pkt_entry;
-
-		if (read_entry->lower_ep_type == SHM_EP) {
-			msg.addr = shm_fiaddr;
-			ret = fi_readmsg(ep->shm_ep, &msg, 0);
-		} else {
-			efa_ep = container_of(ep->rdm_ep, struct efa_ep, util_ep.ep_fid);
-			msg.addr = read_entry->addr;
-			self_comm = (read_entry->context_type == RXR_READ_CONTEXT_PKT_ENTRY);
-			if (self_comm)
-				pkt_entry->flags |= RXR_PKT_ENTRY_LOCAL_READ;
-			ret = efa_rma_post_read(efa_ep, &msg, 0, self_comm);
-		}
-
-		if (OFI_UNLIKELY(ret)) {
-			rxr_pkt_entry_release_tx(ep, pkt_entry);
+			read_once_len = MIN(read_once_len, rxr_env.efa_read_segment_size);
```

Comment:
This is not related to what this PR really does. Just wondering if we can use `max_read_size` here, as L603 already calculate the max read size based on lower ep type.

Reply:
Sure. I think it is good idea. added that in new revision